### PR TITLE
Batch copy: copy permission too

### PR DESCRIPTION
**Pull request by alikon**
follow up  #24730.

### Summary of Changes
copy permission when Batch copy items to a new category

### Testing Instructions

-  Create an article.
-  Go to its permission tab and make a change. (e.g registered->allow creating)
-  Go to Batch and Copy it somewhere else.

### Expected result
the permission are copyied

### Actual result
permission are not copied

### CODE REVIEW

**Comment by richard67** [comment]
I have tested this item :white_check_mark: successfully on c72d6f29333adfb8fcf6d9329852c4b4a0154a2d<br /><br />I've tested with 2 articles which had differenc ACL modifications.
Both articles copied in 1 batch action.
Before patch: Copied articles have ACL like in new articles.
After patch: Each of the article copies has ACL like the particular copy source article.

@HLeithner and @wilsonge Release leads please check as also with PR #24730 if this is a bug fix so it goes into 3.9.next or a new feature for 4.0.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/24736">issues.joomla.org/tracker/joomla-cms/24736</a>.</sub>

**Comment by viocassel** [comment]
I have tested this item :white_check_mark: successfully on c72d6f29333adfb8fcf6d9329852c4b4a0154a2d<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/24736">issues.joomla.org/tracker/joomla-cms/24736</a>.</sub>

**Comment by ghost** [comment]
Status "Ready To Commit".

**Comment by wilsonge** [comment]
I'm not unhappy to call this a bug fix - but what happens if there isn't a entry in the permissions table for the asset - shouldn't we check in JTable that there actually is a asset_id for items

**Comment by infograf768** [comment]
I agree, for this one, with @wilsonge 
I guess we may not need that check for categories and modules.

**Comment by wilsonge** [comment]
Agreed. The stuff that's component specific is fine. But this generic one needs more careful sanity checks

**Comment by alikon** [comment]
not so sure what "sanity check" you are talking  about if  target and/or source asset don't exists the query simply does nothing

**Comment by infograf768** [comment]
I guess just making sure there is an existing asset for the original item.
```
if ($oldAssetId)

query
```

**Comment by infograf768** [comment]
@wilsonge 
the rtc label has been taken off without a reply to @alikon 
Please explain further what you expect to get this in.

File: libraries/src/MVC/Model/AdminModel.php [hunk 1]
```diff
@@ -487,6 +489,18 @@ protected function batchCopy($value, $pks, $contexts)
...
+			// Copy rules
+			$query = $db->getQuery(true);
+			$query->clear()
+				->update($db->quoteName('#__assets', 't'))
+				->join('INNER', $db->quoteName('#__assets', 's') .
+					' ON ' . $db->quoteName('s.id') . ' = ' . $oldAssetId
+				)
+				->set($db->quoteName('t.rules') . ' = ' . $db->quoteName('s.rules'))
+				->where($db->quoteName('t.id') . ' = ' . $this->table->asset_id);
+
```
Comment (Quy) [inline libraries/src/MVC/Model/AdminModel.php]: Please remove tab.

**Comment by HLeithner** [comment]
I removed it because George requested more sanity checks, I'm also not sure if this is a bugfix because permission copying is missing or a new feature.

Atm I think it's a bugfix but ymmv.

**Comment by alikon** [comment]
for sanity check i've already answered https://github.com/joomla/joomla-cms/pull/24736#issuecomment-487354087
:confused:

**Comment by HLeithner** [comment]
@wilsonge could you please elaborate your comment?

**Comment by wilsonge** [comment]
Inside JTable we check for asset support based on the asset id field being present  https://github.com/joomla/joomla-cms/blob/6933bb16b8aa5108a3d9c6aa998f64f4a74e2518/libraries/src/Table/Table.php#L178-L182

we need something similar in the model so that we know if the item actually has an asset or not because right now we're just going to blindly try and clone assets even for items that don't have assets

**Comment by alikon** [comment]
did you mean something like `$this->table->hasField($this->table->getColumnAlias('asset_id'))`
?

**Comment by wilsonge** [comment]
Something like that yes

**Comment by infograf768** [comment]
Back to pending until change.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/24736">issues.joomla.org/tracker/joomla-cms/24736</a>.</sub>

**Comment by alikon** [comment]
change made, please check, retest

**Comment by richard67** [comment]
I'll wait for the travis and appveyor and drone results, then test. Am already preparing.

**Comment by richard67** [comment]
@alikon I get this after having applied your patch on a current staging, clean new install with testing sample datas, using patchtester to apply it:
Unbenannt

**Comment by alikon** [comment]
yes you're right `hasField()`  was introduced in 4.x  backported to 3.x now

**Comment by richard67** [comment]
@alikon Now there is no error, and the articles are copied, but the permissions not.

**Comment by richard67** [comment]
@alikon Sorry, forget previous comment. My mistake.

**Comment by richard67** [comment]
@alikon Hmm, now after having applied the patch, when batch copying 2 articles, only permissions of 1st article are copied.

**Comment by richard67** [comment]
@alikon Sorry, again my mistake. all ok.

**Comment by richard67** [comment]
I have tested this item :white_check_mark: successfully on c0fbbeb1a9c681ac4fe8ff671da9f30d8cfd34dd<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/24736">issues.joomla.org/tracker/joomla-cms/24736</a>.</sub>

File: libraries/src/Table/Table.php [hunk 2]
```diff
@@ -1722,4 +1722,19 @@ protected function _unlock()
...
+	 *
+	 * @param   string  $key  key to be checked
+	 *
+	 * @return  boolean
+	 *
+	 * @since   __DEPLOY_VERSION__
+	 */
+	public function hasField($key)
+	{
+		$key = $this->getColumnAlias($key);
```
Comment (Quy) [inline libraries/src/Table/Table.php:1737]:
```suggestion
		$key = $this->getColumnAlias($key);

```

**Comment by richard67** [comment]
I have tested this item :white_check_mark: successfully on 7b6861a855856c086a8951e93302973f95f32ecb<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/24736">issues.joomla.org/tracker/joomla-cms/24736</a>.</sub>

**Comment by Quy** [comment]
In PHP error log when batching banners.

> PHP Notice:  Undefined property: BannersTableBanner::$asset_id in \joomla-cms-staging\libraries\src\MVC\Model\AdminModel.php on line 439

**Comment by richard67** [comment]
> 
> 
> In PHP error log when batching banners.
> 
> > PHP Notice:  Undefined property: BannersTableBanner::$asset_id in \joomla-cms-staging\libraries\src\MVC\Model\AdminModel.php on line 439

Confirmed.

**Comment by richard67** [comment]
I have not tested this item.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/24736">issues.joomla.org/tracker/joomla-cms/24736</a>.</sub>

File: libraries/src/MVC/Model/AdminModel.php [hunk 3]
```diff
@@ -435,6 +436,15 @@ protected function batchCopy($value, $pks, $contexts)
 				}
 			}
 
+			// Check for asset_id
+			$hasAsset = false;
+
+			if ($this->table->hasField($this->table->getColumnAlias('asset_id')))
```
Comment (wilsonge) [inline libraries/src/MVC/Model/AdminModel.php:440]: Why not just move this down. Put `$oldAssetId` directly in the query and this if statement just down there

**Comment by Quy** [comment]
I have tested this item :white_check_mark: successfully on 49226ba371d9491fe1bf860308494ebfcbe26445<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/24736">issues.joomla.org/tracker/joomla-cms/24736</a>.</sub>

**Comment by ghost** [comment]
@richard67 can you please retest?

**Comment by richard67** [comment]
@franz-wohlkoenig As soon as I have cooked and eaten I will re-test.

**Comment by richard67** [comment]
I have tested this item :white_check_mark: successfully on 49226ba371d9491fe1bf860308494ebfcbe26445<br /><br />1. Permissions are copied.
2. No "PHP Notice: Undefined property ..." anymore when batch copying banners.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/24736">issues.joomla.org/tracker/joomla-cms/24736</a>.</sub>

**Comment by richard67** [comment]
@franz-wohlkoenig Done.

**Comment by ghost** [comment]
Status "Ready To Commit".

**Comment by wilsonge** [comment]
@alikon can you please do the cleanup i mentioned

**Comment by Quy** [comment]
@wilsonge I did PR [#51](https://github.com/alikon/joomla-cms/pull/51/files) as you suggested, but had to revert with PR [#52](https://github.com/alikon/joomla-cms/pull/52/files) due to a warning that I can't remember at the moment.

**Comment by alikon** [comment]
@wilsonge  to me the only thing which i can think about is to move https://github.com/joomla/joomla-cms/blob/7f8bad80600699956e95eb67dacefd5711521b98/libraries/src/MVC/Model/AdminModel.php#L442

outside i.e before the `while`   cause we are always referring to the same table..... 
but `$oldAssetId `  must be stored before `$this->table->asset_id`  is  changed  by `this->table->store()`


or i missunderstand you

**Comment by HLeithner** [comment]
@wilsonge can you comment on this before I merge it?

**Comment by wilsonge** [comment]
Basically what @quy did in https://github.com/alikon/joomla-cms/pull/51/files is what I’m after - it makes more sense if we can make it work. Can we find out why that was failing?

**Comment by richard67** [comment]
@alikon why closing this?

**Comment by richard67** [comment]
@alikon Did you maybe accidently hit the wrong button, close instead of rebase?

**Comment by alikon** [comment]
i've  had 5 minutes of "send all to the hell" feeling...
...a little bit better now

**Comment by richard67** [comment]
@wilsonge The syntax for the join, does the fix for the "ON" which you have done in J4 here [https://github.com/joomla/joomla-cms/pull/25724/commits/a18322dff428bdc4a8d9fbc3ab69738a3ece8684](https://github.com/joomla/joomla-cms/pull/25724/commits/a18322dff428bdc4a8d9fbc3ab69738a3ece8684) apply also to J3? If so, it should be also fixed here in this PR. I will then make PR for Nicola's branch but want you to confirm if that would be right.

**Comment by richard67** [comment]
@wilsonge I just see according to J3 database api doc my previous comment for you is wrong, i.e. this PR here is ok with "... ON ... " in 1 join condition. Is that right?

Regarding your sanity check, Nicola has merged noth PRs from @Quy so now it should be ok. Maybe reason why moving things down did not work is because the `$this->table->get('id');` would be before that then and reset stuff somehow?

File: libraries/src/MVC/Model/AdminModel.php [hunk 4]
```diff
@@ -487,6 +494,21 @@ protected function batchCopy($value, $pks, $contexts)
...
 			$newId = $this->table->get('id');
 
+			if (!empty($oldAssetId))
+			{
+				// Copy rules
+				$query = $db->getQuery(true);
+				$query->clear()
+					->update($db->quoteName('#__assets', 't'))
+					->join('INNER', $db->quoteName('#__assets', 's') .
+						' ON ' . $db->quoteName('s.id') . ' = ' . $oldAssetId
```
Comment (HLeithner) [inline libraries/src/MVC/Model/AdminModel.php:504]: $oldAssetId should be a prepared parameter
```suggestion
						' ON ' . $db->quoteName('s.id') . ' = :oldAssetId1'
```

File: libraries/src/MVC/Model/AdminModel.php [hunk 5]
```diff
@@ -487,6 +494,21 @@ protected function batchCopy($value, $pks, $contexts)
...
+					)
+					->set($db->quoteName('t.rules') . ' = ' . $db->quoteName('s.rules'))
+					->where($db->quoteName('t.id') . ' = ' . $this->table->asset_id);
```
Comment (HLeithner) [inline libraries/src/MVC/Model/AdminModel.php:507]: $this->table->asset_id should be a prepared parameter
```suggestion
					->where($db->quoteName('t.id') . ' = :oldAssetId2')
					->bind([':oldAssetId1', ':oldAssetId2'], $this->table->asset_id, ParameterType::INTEGER);
```

This requires and use statement for ParameterType and you should remove the if statement at line 497 as george suggested.

File: libraries/src/MVC/Model/AdminModel.php [hunk 6]
```diff
@@ -487,6 +494,21 @@ protected function batchCopy($value, $pks, $contexts)
 			// Get the new item ID
 			$newId = $this->table->get('id');
 
+			if (!empty($oldAssetId))
```
Comment (HLeithner) [inline libraries/src/MVC/Model/AdminModel.php:497]:
```suggestion
			if ($this->table->hasField($this->table->getColumnAlias('asset_id')) && !empty($this->table->asset_id))
```

File: libraries/src/MVC/Model/AdminModel.php [hunk 7]
```diff
@@ -487,6 +494,21 @@ protected function batchCopy($value, $pks, $contexts)
...
+			{
+				// Copy rules
```
Comment (HLeithner) [inline libraries/src/MVC/Model/AdminModel.php:499]:
```suggestion
				// Copy rules
				$oldAssetId = $this->table->asset_id;
```

File: libraries/src/MVC/Model/AdminModel.php (hunk 4 above)
Comment (alikon) [reply libraries/src/MVC/Model/AdminModel.php:504]: even in 3.x ?

Comment (HLeithner) [reply libraries/src/MVC/Model/AdminModel.php:504]: ah... sorry you are right, missed that it's J3... forget my comments

**Comment by alikon** [comment]
@HLeithner cause i'm not too much able to read only & fully understand  code
 i've made a test with your suggestion applyied ...... as a result permission are not copyied

so may I ask you to do the same test?
with your suggestion  and without (only this silly pr)
........maybe you'll can confirm or deny my results

File: libraries/src/MVC/Model/AdminModel.php [hunk 8]
```diff
@@ -487,6 +494,21 @@ protected function batchCopy($value, $pks, $contexts)
 			// Get the new item ID
```
Comment (Quy) [inline libraries/src/MVC/Model/AdminModel.php:494]: After this point `$this->table->asset_id` has incremented. Thus the old value must be stored before and not after this point. The PR works as it.

**Comment by richard67** [comment]
@Quy That was what I supposed, too, but seems that comment in code reviews was too hidden.

**Comment by richard67** [comment]
Tonight I can test again. But last changes were code style only so my old test should still be ok.

File: libraries/src/MVC/Model/AdminModel.php (hunk 8 above)
Comment (HLeithner) [reply libraries/src/MVC/Model/AdminModel.php:494]: ok maybe I missed something in the code.

**Comment by richard67** [comment]
I have tested this item :white_check_mark: successfully on 6b24187ecd6f04a36e5f1375d4cf049b7b26d9d9<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/24736">issues.joomla.org/tracker/joomla-cms/24736</a>.</sub>

**Comment by richard67** [comment]
@viocassel or @Quy could you test again? Thanks in advance.

**Comment by viocassel** [comment]
I have tested this item :white_check_mark: successfully on 6b24187ecd6f04a36e5f1375d4cf049b7b26d9d9<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/24736">issues.joomla.org/tracker/joomla-cms/24736</a>.</sub>

**Comment by richard67** [comment]
@viocassel Thanks for testing.

**Comment by ghost** [comment]
Status "Ready To Commit".

**Comment by HLeithner** [comment]
Thank you for your engagement in Joomla!

**Pull request merged**